codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/UnionWithUnitTest.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.server.smithy
+
+import org.junit.jupiter.api.Test
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.testutil.asSmithyModel
+import software.amazon.smithy.rust.codegen.core.testutil.testModule
+import software.amazon.smithy.rust.codegen.core.testutil.unitTest
+import software.amazon.smithy.rust.codegen.server.smithy.testutil.serverIntegrationTest
+
+class UnionWithUnitTest {
+    @Test
+    fun `a constrained union that has a unit member should compile`() {
+        val model =
+            """
+            ${'$'}version: "2"
+            namespace com.example
+            use aws.protocols#restJson1
+            use smithy.framework#ValidationException
+            
+            @restJson1 @title("Test Service") 
+            service TestService { 
+                version: "0.1", 
+                operations: [ 
+                    TestOperation
+                    TestSimpleUnionWithUnit
+                ] 
+            }
+            
+            @http(uri: "/testunit", method: "POST")
+            operation TestSimpleUnionWithUnit {
+                input := {
+                    @required
+                    request: SomeUnionWithUnit
+                }
+                output := {
+                    result : SomeUnionWithUnit
+                }
+                errors: [
+                    ValidationException
+                ]
+            }
+            
+            @length(min: 13)
+            string StringRestricted
+            
+            union SomeUnionWithUnit {
+                Option1: Unit
+                Option2: StringRestricted
+            }
+
+            @http(uri: "/test", method: "POST")
+            operation TestOperation {
+                input := { payload: String }
+                output := {
+                    @httpPayload
+                    events: TestEvent
+                },
+                errors: [ValidationException]
+            }
+            
+            @streaming
+            union TestEvent {
+                KeepAlive: Unit,
+                Response: TestResponseEvent,
+            }
+            
+            structure TestResponseEvent { 
+                data: String 
+            }            
+            """.asSmithyModel()
+
+        serverIntegrationTest(model) { _, rustCrate ->
```

Comment:
Not sure why we need the tests below if all we want to check is that the model compiles correctly.

Reply:
So that we can extend the same test to RpcV2Cbor (and possibly other protocols). Since event streams are not covered by the protocol tests, and none of the other tests were failing, I assume we do not have a test that actually checks whether data is being correctly serialized/deserialized. However, having said that, I think it is better to add comprehensive tests that use a client to serialize and then use the server to deserialize it. Currently, these tests directly use the inner data types/methods rather than invoking the server. Will remove these tests for now and we can add one in v2Cbor.